clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2939,13 +2939,16 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
     return;
 
   WorkGroupAttr *Existing = D->getAttr<WorkGroupAttr>();
-  if (Existing && !(Existing->getXDim() == WGSize[0] &&
-                    Existing->getYDim() == WGSize[1] &&
-                    Existing->getZDim() == WGSize[2]))
+  if (Existing &&
+      !(Existing->getXDim() == WGSize[0] && Existing->getYDim() == WGSize[1] &&
```

Comment:
This check is not reversed. Does the warning below still work correctly?

Reply:
I think it can work. Is there test for it?

Reply:
If yes, then all works fine.

Reply:
> I think it can work. Is there test for it?

That is what I was going to ask.

If I'm reading this code correctly, we compare XDim of an existing attribute with WGSize[0], which is going to be ZDim. Shouldn't we compare existing XDim with WGSize[2] instead?

Reply:
Maybe you are right

Reply:
I will try to prove it.

Reply:
No, it does't need to be reversed. I have tested it.

Reply:
Warning works correctly.

Reply:
Yes, it need to be reversed.

Reply:
I suspect that the diagnostic was not tested. Can you add a test for it?

Reply:
A general case was tested, see: reqd-work-group-size.cpp:23 - 31, and these diagnostics are firing anyway.
But with this patch there is a corner case introduced, that was untested, like:
```
[[cl::reqd_work_group_size(4, 4, 1)]]
[[cl::reqd_work_group_size(1, 1, 4)]]
void foo() { /*...*/ }
```

Reply:
I add one test, that prove rightness of the implementation.